src/ejson2/EjsonEncoder.ts

```diff
@@ -0,0 +1,593 @@
+        this.writeNumber(value.value);
+      }
+    }
+  }
+
+  private writeBsonDecimal128AsEjson(value: BsonDecimal128): void {
+    // Write {"$numberDecimal":"..."}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x22246e75); writer.u32(0x6d626572); writer.u32(0x44656369); writer.u32(0x6d616c22); // "$numberDecimal"
+    writer.u8(0x3a); // :
+    this.writeStr(this.decimal128ToString(value.data));
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonBinaryAsEjson(value: BsonBinary): void {
+    // Write {"$binary":{"base64":"...","subType":"..."}}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x22246269); writer.u32(0x6e617279); writer.u8(0x22); // "$binary"
+    writer.u8(0x3a); // :
+    writer.u8(0x7b); // {
+    writer.u32(0x22626173); writer.u32(0x65363422); // "base64"
+    writer.u8(0x3a); // :
+    this.writeStr(this.uint8ArrayToBase64(value.data));
+    writer.u8(0x2c); // ,
+    writer.u32(0x22737562); writer.u32(0x54797065); writer.u8(0x22); // "subType"
+    writer.u8(0x3a); // :
+    this.writeStr(value.subtype.toString(16).padStart(2, '0'));
+    writer.u8(0x7d); // }
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonCodeAsEjson(value: BsonJavascriptCode): void {
+    // Write {"$code":"..."}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x2224636f); writer.u16(0x6465); writer.u8(0x22); // "$code"
+    writer.u8(0x3a); // :
+    this.writeStr(value.code);
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonCodeWScopeAsEjson(value: BsonJavascriptCodeWithScope): void {
+    // Write {"$code":"...","$scope":{...}}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x2224636f); writer.u16(0x6465); writer.u8(0x22); // "$code"
+    writer.u8(0x3a); // :
+    this.writeStr(value.code);
+    writer.u8(0x2c); // ,
+    writer.u32(0x22247363); writer.u32(0x6f706522); // "$scope"
+    writer.u8(0x3a); // :
+    this.writeAny(value.scope);
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonSymbolAsEjson(value: BsonSymbol): void {
+    // Write {"$symbol":"..."}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x22247379); writer.u32(0x6d626f6c); writer.u8(0x22); // "$symbol"
+    writer.u8(0x3a); // :
+    this.writeStr(value.symbol);
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonTimestampAsEjson(value: BsonTimestamp): void {
+    // Write {"$timestamp":{"t":...,"i":...}}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x22247469); writer.u32(0x6d657374); writer.u32(0x616d7022); // "$timestamp"
+    writer.u8(0x3a); // :
+    writer.u8(0x7b); // {
+    writer.u16(0x2274); writer.u8(0x22); // "t"
+    writer.u8(0x3a); // :
+    this.writeNumber(value.timestamp);
+    writer.u8(0x2c); // ,
+    writer.u16(0x2269); writer.u8(0x22); // "i"
+    writer.u8(0x3a); // :
+    this.writeNumber(value.increment);
+    writer.u8(0x7d); // }
+    writer.u8(0x7d); // }
```

Comment:
Combine every two u8() calls into a single u16() call. And u16 into u32. Everywhere in this file.